presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestPushDistinctLimitIntoTableScan.java

```diff
@@ -0,0 +1,180 @@
+public class TestPushDistinctLimitIntoTableScan
+        extends BaseRuleTest
+{
+    private static final CatalogName TEST_CATALOG = new CatalogName("test_push_dl_catalog");
+
+    private PushDistinctLimitIntoTableScan rule;
+    private TableHandle tableHandle;
+
+    private ApplyAggregation testApplyAggregation;
+
+    @Override
+    protected Optional<LocalQueryRunner> createLocalQueryRunner()
+    {
+        Session defaultSession = TestingSession.testSessionBuilder()
+                .setCatalog(TEST_CATALOG.getCatalogName())
+                .setSchema("tiny")
+                .build();
+
+        LocalQueryRunner queryRunner = LocalQueryRunner.create(defaultSession);
+
+        queryRunner.createCatalog(
+                TEST_CATALOG.getCatalogName(),
+                MockConnectorFactory.builder()
+                        .withApplyAggregation(
+                                (session, handle, aggregates, assignments, groupingSets) -> {
+                                    if (testApplyAggregation != null) {
+                                        return testApplyAggregation.apply(session, handle, aggregates, assignments, groupingSets);
+                                    }
+                                    return Optional.empty();
+                                })
+                        .build(),
+                Map.of());
+
+        return Optional.of(queryRunner);
+    }
+
+    @BeforeClass
+    public void init()
+    {
+        rule = new PushDistinctLimitIntoTableScan(tester().getMetadata());
+
+        tableHandle = new TableHandle(
+                TEST_CATALOG,
+                new MockConnectorTableHandle(new SchemaTableName("mock_schema", "mock_nation")),
+                MockConnectorTransactionHandle.INSTANCE,
+                Optional.empty());
+    }
+
+    @BeforeMethod
+    public void reset()
+    {
+        testApplyAggregation = null;
+    }
+
+    @Test
+    public void testDoesNotFireIfNoTableScan()
+    {
+        tester().assertThat(rule)
+                .on(p -> p.values(p.symbol("a", BIGINT)))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testNoEffect()
+    {
+        AtomicInteger applyCallCounter = new AtomicInteger();
+        testApplyAggregation = (session, handle, aggregates, assignments, groupingSets) -> {
+            applyCallCounter.incrementAndGet();
+            return Optional.empty();
+        };
+
+        tester().assertThat(rule)
+                .on(p -> {
+                    Symbol regionkey = p.symbol("regionkey");
+                    return p.distinctLimit(10, List.of(regionkey),
+                            p.tableScan(
+                                    tableHandle,
+                                    ImmutableList.of(regionkey),
+                                    ImmutableMap.of(regionkey, new MockConnectorColumnHandle("regionkey", BIGINT))));
+                })
+                .doesNotFire();
```

Comment:
nit: `doesNotFire` is kinda misleading here. As rule is fired. Yet it returns `empty()` so it does nothing.

Reply:
pre existing,

Reply:
> empty() so it does nothing.

A rule returning `empty()` means "it didn't fire". Firing means that the rule resulted in some transformation of the plan.

Reply:
Before we had patterns -- the meaning was obvious.
With patterns -- the naming is no longer obvious, and i see @losipiuk 's point.
If you remember that pattern is an optional thing and rule may or may not use it (kind of), it becomes clearer -- but there is no way to get this thinking from the code. 

I am OK with changing the method name if we can propose something better, but that would fall out of scope of this PR anyway.

Reply:
Patterns or no patterns, it's the same. Some things cannot be represented by patterns and need to matched manually. The pattern is just a convenience mechanism, but it doesn't mean that if the pattern matches the rule "fires".

I thing you're confusing "fires" with "is invoked".

Reply:
In JMX stats for optimizer rules we use "hits" and "misses" counter. Fire does mean hit.

Reply:
Yes, fire means hit. Any time a rule returns non-empty, we tick the "hits" counter. Otherwise, we tick the "misses" counter. The "invocations" counter gets ticked in either case.

Reply:
> I thing you're confusing "fires" with "is invoked".

i am under impression in some contexts (not here), these are used interchangeably.

Perhaps replacing  "fires" with "applies" would provide for more obvious terminology.

Reply:
> i am under impression in some contexts (not here), these are used interchangeably.

Can you give an example? I'm not aware of any such cases.

Reply:
I meant spoken "dev english", not the Presto code base.

Reply:
>> Fire does not mean hit.
> Yes, fire means hit

Surely, I am not an english speaker. But if a bullet is fired (launched), it does mean it will hit the target. Hence "fire" is closer for me in the meaning to "invocations". But I get used to it the current Presto glossary.